### PR TITLE
Raven db 13796

### PR DESCRIPTION
**Pull request by maximburyak**
This PR was initially meant to investigate a failing test (RachisTests.SubscriptionsFailover.SetMentorToSubscriptionWithFailover), but finished up resolving a race condition between a request executor's topology update and a command's failover

### CODE REVIEW

File: src/Raven.Client/Http/NodeSelector.cs [hunk 1]
```diff
@@ -100,6 +100,23 @@ public bool OnUpdateTopology(Topology topology, bool forceUpdate = false)
             return UnlikelyEveryoneFaultedChoice(state);
         }
 
+        public (int Index, ServerNode Node, long TopologyEtag) GetPreferredNodeWithTopology()
```
Comment (ayende) [inline src/Raven.Client/Http/NodeSelector.cs]: Isn't this duplicating the code in `GetPreferredNode`? Can we merge these?

File: src/Raven.Client/Http/RavenCommand.cs [hunk 2]
```diff
@@ -51,6 +51,8 @@ public abstract class RavenCommand<TResult>
         public bool CanCacheAggressively { get; protected set; }
         public string SelectedNodeTag { get; protected set; }
 
+        public long InitialTopologyEtag = -1;
```
Comment (ppekrol) [inline src/Raven.Client/Http/RavenCommand.cs]: internal

**Comment by maximburyak** [comment]
*Looking into failed tests

File: src/Raven.Client/Http/NodeSelector.cs (hunk 1 above)
Comment (maximburyak) [reply src/Raven.Client/Http/NodeSelector.cs]: done

File: src/Raven.Client/Http/RavenCommand.cs [hunk 3]
```diff
@@ -51,6 +51,8 @@ public abstract class RavenCommand<TResult>
         public bool CanCacheAggressively { get; protected set; }
         public string SelectedNodeTag { get; protected set; }
 
+        internal long FailoverTopologyEtag = -2;
```
Comment (ppekrol) [inline src/Raven.Client/Http/RavenCommand.cs:54]: maybe make it nullable?

File: src/Raven.Client/Http/RequestExecutor.cs [hunk 4]
```diff
@@ -647,6 +647,10 @@ protected virtual bool TryLoadFromCache(JsonOperationContext context)
             SessionInfo sessionInfo = null,
             CancellationToken token = default)
         {
+            if (command.FailoverTopologyEtag == -2)
+            {                
+                command.FailoverTopologyEtag = _nodeSelector?.GetPreferredNodeWithTopology().TopologyEtag ?? -2;
```
Comment (ppekrol) [inline src/Raven.Client/Http/RequestExecutor.cs]: isn't this the same as `RequestExecutor.TopologyEtag` property?

Comment (ayende) [reply src/Raven.Client/Http/RequestExecutor.cs]: That is not the same, no. Because the topology may change *while* the command is running

Comment (ppekrol) [reply src/Raven.Client/Http/RequestExecutor.cs]: The source of it is the same. `NodeSelector.Topology` which is taken from `State` inside the NodeSelector.

**Review by ppekrol — APPROVED** [review]

**Pull request merged**